### PR TITLE
[AI] Speed up TurboQuant Hadamard rotation by ~3x

### PR DESCRIPTION
**Pull request by agourlay**
## What

`HadamardRotation::apply`/`apply_inverse` are on the TurboQuant encode and decode hot paths (every vector goes through them at encoding time and at rescoring/decode time). 

Profiling `apply_inverse` showed only 23% of cycles in the WHT kernel itself; 44% went to replaying the Fisher-Yates shuffle on every call (one serial LCG step plus a 64-bit `%` by a variable bound per element), and another ~12% to the separate normalization pass and buffer copies.

## How

1. **Materialized permutation maps.** `HadamardRotation::new` now builds forward/backward index maps (`[Vec<u32>; 3]` each, 24 bytes per dimension) once, so each permutation at apply time is a flat gather pass instead of a serial swap replay. The gathers ping-pong between the vector and a thread-local scratch buffer (no per-call allocation), with a single copy back at the end.
2. **Fused WHT outer stages (AVX2).** Consecutive butterfly stage pairs (h, 2h) are applied in one pass over the array instead of two, halving memory traffic for all stages h >= 16.
3. **Fused normalization.** `wht_dispatch_scaled` multiplies by the chunk norm in the transform's final-stage stores (via `const SCALED: bool` threaded through the AVX2 kernels), removing the separate read-modify-write normalize pass.
4. The old replay-based implementation (`random_vector_rotation*`, `Permutation::new_reversible`/`unpermute`, reverse LCG) is kept under `#[cfg(test)]` as the parity oracle, leaving a single production path.

## Bit-equality

The permutation seeds are baked into every stored quantized vector, so output must not change by even one ulp. 

All three changes preserve exact op sequences: maps move the same values to the same positions, fused butterflies are the same single add/sub per pair, and the fused norm is still "final add/sub, then one multiply" per element. This is enforced by:

* `static_rotation_matches_struct_and_roundtrips` pins the new map-based path bit-for-bit to the historical replay implementation.
* The existing SIMD-vs-scalar bit-equal tests (all sizes 1 to 8192, random and edge-value inputs).
* New `wht_dispatch_scaled_matches_scalar_then_multiply` covers every dispatch path (n < 8, 8, 16, 32, >= 64) with scales from 1e-160 to 1e160.

## Results

Criterion `hadamard` bench, Zen 5, paired same-session runs (2.8-3.7x):

| dim | `apply` | `apply_inverse` |
|---|---|---|
| 128 | 820ns -> 247ns (3.3x) | 819ns -> 236ns (3.5x) |
| 384 | 2.58µs -> 0.70µs (3.7x) | 2.26µs -> 0.70µs (3.2x) |
| 768 | 5.32µs -> 1.44µs (3.7x) | 4.45µs -> 1.47µs (3.0x) |
| 1024 | 6.98µs -> 2.03µs (3.4x) | 6.10µs -> 2.05µs (3.0x) |
| 1536 | 10.14µs -> 2.97µs (3.4x) | 9.01µs -> 3.08µs (2.9x) |
| 4096 | 27.2µs -> 9.19µs (3.0x) | 25.2µs -> 9.10µs (2.8x) |

## Notes

* NEON is unchanged: it keeps the per-stage outer loop and the separate normalize pass, byte-for-byte the previous behavior (could not benchmark ARM locally). Porting the fused stages and norm fusion is a possible follow-up.
* `HadamardRotation::new` asserts `dim <= u32::MAX` (index maps store u32), and the unchecked gather hard-asserts length equality, so API misuse panics rather than reading out of bounds.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/quantization/src/turboquant/rotation.rs [hunk 1]
```diff
@@ -33,12 +68,69 @@ impl HadamardRotation {
...
     pub fn apply(&self, x: &mut [f64]) {
         debug_assert_eq!(x.len(), self.dim);
-        apply_rotation_with_permutations(x, &self.permutations);
+        wht_and_gather_rounds(x, self.forward_maps.iter());
     }
 
     pub fn apply_inverse(&self, y: &mut [f64]) {
         debug_assert_eq!(y.len(), self.dim);
-        apply_inverse_rotation_with_permutations(y, &self.permutations);
+        wht_and_gather_rounds(y, self.backward_maps.iter().rev());
```
[the remark this answers is not in the document]
Comment (agourlay) [reply lib/quantization/src/turboquant/rotation.rs:78]: Addressed in b4331635a: both boundary checks upgraded to hard `assert_eq!` so a length mismatch panics before any mutation.

**Review by JojiiOfficial — APPROVED** [review]
I don't get 3x on my machine but still a big improvement over `dev` (results below).

Dev and PR are bit compatible as long as the vector doesn't contain `inf` dimensions (I think it's fine). The code itself seems also ok. Approving.


My results:
```

┌────────────────────────┬──────────┬──────────┬─────────┐
│          Dim           │   dev    │    PR    │ Speedup │
├────────────────────────┼──────────┼──────────┼─────────┤
│ hadamard_apply         │          │          │         │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 128                    │ 905.8 ns │ 382.9 ns │ 2.37×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 384                    │ 2.99 µs  │ 1.07 µs  │ 2.79×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 768                    │ 5.92 µs  │ 2.30 µs  │ 2.57×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 1024                   │ 8.19 µs  │ 3.07 µs  │ 2.67×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 1536                   │ 11.99 µs │ 4.63 µs  │ 2.59×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 4096                   │ 34.15 µs │ 15.23 µs │ 2.24×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ hadamard_apply_inverse │          │          │         │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 128                    │ 894.7 ns │ 375.4 ns │ 2.38×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 384                    │ 2.67 µs  │ 1.08 µs  │ 2.47×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 768                    │ 5.56 µs  │ 2.31 µs  │ 2.41×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 1024                   │ 7.84 µs  │ 3.26 µs  │ 2.40×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 1536                   │ 11.65 µs │ 5.00 µs  │ 2.33×   │
├────────────────────────┼──────────┼──────────┼─────────┤
│ 4096                   │ 33.77 µs │ 15.07 µs │ 2.24×   │
└────────────────────────┴──────────┴──────────┴─────────┘
```

**Review by IvanPleshkov — APPROVED** [review]
